numpyro/examples/vae.py

```diff
@@ -0,0 +1,152 @@
+
+
+Sigmoid = _elemwise_no_params(sigmoid)
+
+
+RESULTS_DIR = os.path.abspath(os.path.join(os.path.dirname(__file__),
+                              '.results'))
+os.makedirs(RESULTS_DIR, exist_ok=True)
+
+
+def encoder(hidden_dim, z_dim):
+    return stax.serial(
+        stax.Dense(hidden_dim, W_init=stax.randn()), stax.Softplus,
+        stax.FanOut(2),
+        stax.parallel(stax.Dense(z_dim, W_init=stax.randn()),
+                      stax.serial(stax.Dense(z_dim, W_init=stax.randn()), stax.Exp)),
+    )
+
+
+def decoder(hidden_dim, out_dim):
+    return stax.serial(
+        stax.Dense(hidden_dim, W_init=stax.randn()), stax.Softplus,
+        stax.Dense(out_dim, W_init=stax.randn()), Sigmoid,
+    )
+
+
+def model(batch, **kwargs):
+    decode = kwargs['decode']
+    decoder_params = param('decoder', None)
+    z_dim = kwargs['z_dim']
+    batch = np.reshape(batch, (batch.shape[0], -1))
+    z = sample('z', dist.norm(np.zeros((z_dim,)), np.ones((z_dim,))))
+    img_loc = decode(decoder_params, z)
+    return sample('obs', dist.bernoulli(img_loc), obs=batch)
+
+
+def guide(batch, **kwargs):
+    encode = kwargs['encode']
+    encoder_params = param('encoder', None)
+    batch = np.reshape(batch, (batch.shape[0], -1))
+    z_loc, z_std = encode(encoder_params, batch)
+    z = sample('z', dist.norm(z_loc, z_std))
+    return z
+
+
+@jit
+def binarize(rng, batch):
+    rng, = random.split(rng, 1)
+    return rng, random.bernoulli(rng, batch).astype(batch.dtype)
+
+
+def main(args):
+    encoder_init, encode = encoder(args.hidden_dim, args.z_dim)
+    decoder_init, decode = decoder(args.hidden_dim, 28 * 28)
+    opt_init, opt_update = optimizers.adam(args.learning_rate)
+    svi_init, svi_update, svi_eval = svi(model, guide, elbo, opt_init, opt_update,
+                                         encode=encode, decode=decode, z_dim=args.z_dim)
+    svi_update = jit(svi_update)
+    rng = PRNGKey(0)
+    train_init, train_fetch = load_dataset(MNIST, batch_size=args.batch_size, split='train')
+    test_init, test_fetch = load_dataset(MNIST, batch_size=args.batch_size, split='test')
+    num_train, train_idx = train_init()
+    _, encoder_params = encoder_init((args.batch_size, 28 * 28))
+    _, decoder_params = decoder_init((args.batch_size, args.z_dim))
+    params = {'encoder': encoder_params, 'decoder': decoder_params}
+    rng, sample_batch = binarize(rng, train_fetch(0, train_idx)[0])
+    opt_state = svi_init(rng, (sample_batch,), (sample_batch,), params)
+    rng, = random.split(rng, 1)
+
+    @jit
+    def epoch_train(opt_state, rng):
+        def body_fn(i, val):
+            loss_sum, opt_state, rng = val
+            rng, batch = binarize(rng, train_fetch(i, train_idx)[0])
+            loss, opt_state, rng = svi_update(i, opt_state, rng, (batch,), (batch,),)
+            loss_sum += loss
+            return loss_sum, opt_state, rng
+
+        return lax.fori_loop(0, num_train, body_fn, (0., opt_state, rng))
```

Comment:
Even here jitting the for loop brings down the runtime to 25% of just jitting `body_fn`. Thanks for the tip, @fehiepsi.